contrib/mypy/src/python/pants/contrib/mypy/tasks/mypy_task.py

```diff
@@ -60,9 +82,30 @@ def is_non_synthetic_python_target(target):
   def is_python_target(target):
     return isinstance(target, PythonTarget)
 
-  def _calculate_python_sources(self, targets):
-    """Generate a set of source files from the given targets."""
-    python_eval_targets = filter(self.is_non_synthetic_python_target, targets)
+  def _is_tagged_target(self, target: Target) -> bool:
+    return self.get_options().whitelist_tag_name in target.tags
+
+  def _is_tagged_non_synthetic_python_target(self, target: Target) -> bool:
+    return (self.is_non_synthetic_python_target(target) and
+            self._is_tagged_target(target))
+
+  def _all_targets_are_whitelisted(self, targets: List[Target], all_targets: List[Target]) -> bool:
+    return len(targets) == 0 or len(targets) == len(all_targets)
+
+  def _whitelist_warning(self) -> None:
+    self.context.log.warn(self.WARNING_MESSAGE)
+
+  def _calculate_python_sources(self, target_roots: List[Target]):
+    """Filter targets to generate a set of source files from the given targets."""
+    if self.get_options().whitelist_tag_name:
+      white_listed_targets = self._filter_targets(Target.closure_for_targets(list(filter(self._is_tagged_target, target_roots))))
+      python_eval_targets = [tgt for tgt in list(white_listed_targets) if self._is_tagged_non_synthetic_python_target(tgt)]
+      if not self._all_targets_in_context_are_whitelisted(python_eval_targets, white_listed_targets):
+        self._whitelist_warning()
+    else:
+      python_eval_targets = self._filter_targets(list(filter(self.is_non_synthetic_python_target, Target.closure_for_targets(target_roots))))
+      self._whitelist_warning()
```

Comment:
I don't think that we want this warning to go off if they choose to not use the whitelist mechanism. We shouldn't make people use a whitelist. If they are using a whitelist, then I agree the warning is useful though.